server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistEventsStep.java

```diff
@@ -0,0 +1,129 @@
+import org.sonar.server.db.DbClient;
+
+import java.util.List;
+
+public class PersistEventsStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final System2 system2;
+
+  public PersistEventsStep(DbClient dbClient, System2 system2) {
+    this.dbClient = dbClient;
+    this.system2 = system2;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext context) {
+    DbSession session = dbClient.openSession(false);
+    try {
+      int rootComponentRef = context.getReportMetadata().getRootComponentRef();
+      recursivelyProcessComponent(session, context, rootComponentRef);
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private void recursivelyProcessComponent(DbSession session, ComputationContext context, int componentRef) {
+    BatchReportReader reportReader = context.getReportReader();
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    processEvents(session, component, context.getReportMetadata().getAnalysisDate());
+    createVersionEvent(session, component, context.getReportMetadata().getAnalysisDate());
+
+    for (Integer childRef : component.getChildRefsList()) {
+      recursivelyProcessComponent(session, context, childRef);
+    }
+  }
+
+  private void processEvents(DbSession session, BatchReport.Component component, Long analysisDate) {
+    List<BatchReport.Event> events = component.getEventsList();
+    if (!events.isEmpty()) {
+      for (BatchReport.Event event : component.getEventsList()) {
+        dbClient.eventDao().insert(session, createBaseEvent(component, analysisDate)
+            .setName(event.getName())
+            .setCategory(convertCategory(event.getCategory()))
+            .setDescription(event.hasDescription() ? event.getDescription() : null)
+            .setData(event.hasEventData() ? event.getEventData() : null)
+          );
+      }
+    }
+  }
+
+  private void createVersionEvent(DbSession session, BatchReport.Component component, Long analysisDate) {
+    if (component.hasVersion()) {
+      deletePreviousEventsHavingSameVersion(session, component);
+      dbClient.eventDao().insert(session, createBaseEvent(component, analysisDate)
+          .setName(component.getVersion())
+          .setCategory(EventDto.CATEGORY_VERSION)
+        );
+    }
+  }
+
+  private void deletePreviousEventsHavingSameVersion(DbSession session, BatchReport.Component component){
+    for (EventDto dto : dbClient.eventDao().selectByComponentUuid(session, component.getUuid())) {
+      if (dto.getCategory().equals(EventDto.CATEGORY_VERSION) && dto.getName().equals(component.getVersion())) {
+        dbClient.eventDao().delete(session, dto.getId());
+      }
+    }
+  }
+
+  private EventDto createBaseEvent(BatchReport.Component component, Long analysisDate) {
```

Comment:
Ok, I'll rename createVersionEvent to saveVersionEvent and rename createBaseEvent to newBaseEvent